### PR TITLE
wgpu

### PR DESCRIPTION
**Pull request by jazzfool**

### CODE REVIEW

File: src/graphics/canvas.rs [hunk 1]
```diff
@@ -1,392 +1,85 @@
...
-        }
-        None => {
-            ctx.gfx_context.data.out = ctx.gfx_context.screen_render_target.clone();
-        }
-    };
+/// Describes the image load operation when starting a new canvas.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CanvasLoadOp {
+    /// Keep the existing contents of the image.
+    DontClear,
```
Comment (aleokdev) [inline src/graphics/canvas.rs:427]: DoNotClear is maybe better, Dont sounds too informal

Comment (jazzfool) [reply src/graphics/canvas.rs:427]: i think this language is ok, vulkan for example has `VK_ATTACHMENT_LOAD_OP_DONT_CARE`

File: README.md [hunk 2]
```diff
@@ -1,188 +1,40 @@
...
+The following platforms receive primary support and are known to work:
 
- If you still have problems or questions, feel free to ask!  Easiest ways are:
+- Window
+- Linux
+- MacOS
 
- * Open an issue on [the Github issue tracker](https://github.com/ggez/ggez/issues)
- * Say hi on the [unofficial Rust Discord server](http://bit.ly/rust-community), the [Rust Gamedev server](https://discord.gg/yNtPTb2) or the [good-web-game Discord server](https://discord.gg/jum3Fjek2A)
+The following platforms are untested, but the backend is compatible (i.e. no major changes needed to make fully compatible):
```
Comment (aleokdev) [inline README.md]: What about web?

Comment (jazzfool) [reply README.md]: ah yeah, web too

File: src/error.rs [hunk 3]
```diff
@@ -5,7 +5,7 @@ use std::sync::Arc;
 use thiserror::Error;
 
 /// An enum containing all kinds of game framework errors.
-#[derive(Error, Debug, Clone)]
+#[derive(Error, Debug)]
```
Comment (aleokdev) [inline src/error.rs]: No clone?

File: src/error.rs [hunk 4]
```diff
@@ -54,6 +54,9 @@ pub enum GameError {
     /// Something went wrong with the `lyon` shape-tesselation library.
     #[error("lyon tesellation error: {0}")]
     LyonError(String),
+    /// Something went wrong when spawing a text with `futures`.
```
Comment (aleokdev) [inline src/error.rs]: spawning*

File: src/graphics/canvas.rs [hunk 5]
```diff
@@ -538,8 +565,105 @@ impl<'a> Canvas<'a> {
...
+                ..
+            } = self;
+
+            std::mem::drop(pass);
+
+            glyph_brush
+                .draw_queued_with_transform(
+                    device,
+                    staging_belt,
+                    unsafe { encoder.as_mut().unwrap() },
```
Comment (aleokdev) [inline src/graphics/canvas.rs]: Missing safety guarantee

File: src/error.rs (hunk 3 above)
Comment (jazzfool) [reply src/error.rs]: spawnerror
might `Arc` it later to restore clone tho

File: src/graphics/canvas.rs (hunk 5 above)
Comment (jazzfool) [reply src/graphics/canvas.rs]: pass is the only thing borrowing encoder so that's explicitly dropped above
quite harmless anywho

**Review by aleokdev — APPROVED** [review]

File: src/audio.rs [hunk 6]
```diff
@@ -85,9 +81,9 @@ pub struct SoundData(Arc<[u8]>);
 
 impl SoundData {
     /// Load the file at the given path and create a new `SoundData` from it.
-    pub fn new<P: AsRef<path::Path>>(context: &mut Context, path: P) -> GameResult<Self> {
+    pub fn new<P: AsRef<path::Path>>(fs: &Filesystem, path: P) -> GameResult<Self> {
```
Comment (aleokdev) [inline src/audio.rs:66]: At this point this PR is really just changing everything

**Comment by aleokdev** [comment]
Should we refcount image? wasn't it refcounted before? Canvas uses a ref to it instead of an arc

**Comment by PSteinhaus** [comment]
I'm still trying to read and understand more of your fork. Until I get a deeper understanding I can only be of limited help and point you towards tiny things, such as:

- Why do you use enumerate here, if you don't use the enumeration? It can just be removed, right? https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/gpu/growing.rs#L27
- These comments should probably refer to text, not meshes, right? https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L282 https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L296 https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L310

**Comment by jazzfool** [comment]
TODO list before this can be considered for a PR
- [x] Reach API parity with `ggez::graphics`
- [x] Transition all examples
- [x] Publicly expose WGPU types (allows for user integration of other rendering, e.g. egui)
    - Also involves exposing the current frame's `CommandEncoder`
- [x] Stress test *all* parts of the new API to find bugs
    - [x] Growing uniform buffer
    - [x] MSAA canvases
    - [x] Instance arrays
    - [x] Custom shaders
    - [x] Custom shader uniforms, textures, and samplers
- [x] ? Z-sorted queue for `Canvas`
- [ ] ~~? Test web support~~ out of scope for PR

**Comment by aleokdev** [comment]
Refer to the upstream PR

**Pull request closed without merging**